src/zimscraperlib/executor.py

```diff
@@ -0,0 +1,179 @@
+            if self.no_more:
+                raise RuntimeError(
+                    "cannot submit task to a joined executor, restart it first"
+                )
+            if _shutdown:
+                raise RuntimeError(  # pragma: no cover
+                    "cannot submit task after interpreter shutdown"
+                )
+
+        while True:
+            try:
+                self.put((task, kwargs), block=True, timeout=3.0)
+            except queue.Full:
+                if self.no_more:
+                    # rarely happens except if submit and join are done in different
+                    # threads, but we need this to escape the while loop
+                    break  # pragma: no cover
+            else:
+                break
+
+    def start(self):
+        """Enable executor, starting requested amount of workers
+
+        Workers are started always, not provisioned dynamically"""
+        logger.debug(f"Starting {self.executor_name} with {self.nb_workers} threads")
+
+        self.drain()
+        self._workers: set[threading.Thread] = set()
+        self.no_more = False
+        self._shutdown = False
+        self.exceptions[:] = []
+
+        for n in range(self.nb_workers):
+            t = threading.Thread(target=self.worker, name=f"{self.executor_name}-{n}")
+            t.daemon = True
+            t.start()
+            self._workers.add(t)
+
+    def worker(self):
+        while self.alive or self.no_more:
+            try:
+                func, kwargs = self.get(block=True, timeout=2.0)
+            except queue.Empty:
+                if self.no_more:
+                    break
+                continue
+            except TypeError:  # pragma: no cover
+                # received None from the queue. most likely shuting down
+                return
+
+            raises = kwargs.pop("raises") if "raises" in kwargs.keys() else False
+            callback = kwargs.pop("callback") if "callback" in kwargs.keys() else None
+            dont_release = kwargs.pop("dont_release", False)
+
+            try:
+                func(**kwargs)
+            except Exception as exc:
+                logger.error(f"Error processing {func} with {kwargs=}")
+                logger.exception(exc)
+                if raises:  # to cover when raises = False
+                    self.exceptions.append(exc)
+                    self.shutdown()
+            finally:
+                # user will manually release the queue for this task.
+                # most likely in a libzim-written callback
+                if not dont_release:
+                    self.task_done()
+                if callback:
+                    callback.__call__()
+
+    def drain(self):
+        """Empty the queue without processing the tasks (tasks will be lost)"""
+        while True:
+            try:
+                self.get_nowait()
+            except queue.Empty:
+                break
+
+    def join(self):
```

Comment:
What's the recommended way to await an executor completion then?
With this, the `deadline_sec` being mandatory, I can only use `join` when I want to exit but in a scraper, I imagine the scenario being: I collect all my resources and submit them to the executor then I join and once everything has been processed (once join completes), I can continue.

This is the regular meaning of `join`.

If we want it to act as a _properly exit_ method, then the user has to track progress manually and this should be clearly specified in the documentation of the executor